compiler/rustc_hir_analysis/src/check/compare_method.rs

```diff
@@ -555,19 +555,25 @@ pub fn collect_trait_impl_trait_tys<'tcx>(
                     // contains `def_id`'s early-bound regions.
                     let id_substs = InternalSubsts::identity_for_item(tcx, def_id);
                     debug!(?id_substs, ?substs);
-                    let map: FxHashMap<ty::GenericArg<'tcx>, ty::GenericArg<'tcx>> = substs
-                        .iter()
-                        .enumerate()
-                        .map(|(index, arg)| (arg, id_substs[index]))
-                        .collect();
+                    let map: FxHashMap<ty::GenericArg<'tcx>, ty::GenericArg<'tcx>> =
+                        std::iter::zip(substs, id_substs).collect();
                     debug!(?map);
 
+                    // HACK(RPITIT): We want this type's regions to act like early bound regions
+                    // on the impl's method generics, not the trait's method generics, so adjust
+                    // the index of each early bound region to account for the fact that the trait
+                    // and impl method substs have different lengths.
                     let ty = tcx.fold_regions(ty, |region, _| {
-                        if let ty::ReFree(_) = region.kind() {
-                            map[&region.into()].expect_region()
-                        } else {
-                            region
-                        }
+                        let ty::ReFree(_) = region.kind() else { return region; };
+                        let ty::ReEarlyBound(e) = map[&region.into()].expect_region().kind()
+                            else { bug!("expected ReFree to map to ReEarlyBound"); };
+                        tcx.mk_region(ty::ReEarlyBound(ty::EarlyBoundRegion {
+                            def_id: e.def_id,
+                            name: e.name,
+                            index: (e.index as usize - trait_to_impl_substs.len()
+                                + tcx.generics_of(impl_m.container_id(tcx)).params.len())
+                                as u32,
```

Comment:
`Generics::param_def_id_to_index` should give the correct answer.  Is `e.def_id` the wrong definition to refer to?

Reply:
Yes, `e.def_id`  points to the RPIT lifetime on the trait. There is no corresponding RPIT generics for the impl.

Reply:
I see 3 cases:
- there exists some early-bound lifetime on the impl that we could map to;
- the corresponding lifetime in the impl is late-bound, and we should build the correct binder;
- there is no lifetime because there is no unification, and we should probably return `re_static`.

Reply:
@cjgillot:

> the corresponding lifetime in the impl is late-bound, and we should build the correct binder

I'm not really sure what you mean by this. The inferred opaque type shouldn't have any late-bound regions -- associated types don't have late-bound regions, do they? For example, let's look at a manual desugaring of an RPITIT:

```
trait Foo {
  fn foo<'a>(&'a self) -> impl Display + 'a;
}

struct Wrapper<T>(T);

impl<T: Display> Foo for Wrapper<T> {
  fn foo<'a>(&'a self) -> impl Display + 'a {
    &self.0
  }
}
```

```
trait Foo {
  type FooRpit<'a>: Display + 'a;

  fn foo<'a>(&'a self) -> &'a T;
}

struct Wrapper<T>(T);

impl Foo for Wrapper<i32> {
  type FooRpit<'a> = &'a i32;

  fn foo<'a>(&'a self) -> Self::FooRpit<'a> {
    &self.0
  }
}
```

Since `{impl#0}::FooRpit` isn't actually a real item, during the RPIT inference step, we need to "rebase" the opaque type's regions (from the trait) onto the impl's own substs as if that item existed, so we can later substitute things correctly during projection. Since the `impl` has no substs of its own, the early bound regions of `FooRpit` start counting from zero, not one.

If we wrapped things in binders instead, then I think we'd be treating the function `foo` in the `impl` like `for<'a> fn(&'a self) -> for<'b> &'b i32` even though the lifetime of `&self` and `&i32` in that above example should be related through substs.

Reply:
Thank you for the detailed explanation.  This is definitely subtle.

Do you mind adding a comment in the code to explain that we are working around the non-existence of `{impl#0}::FooRpit`?

You shift the subst index by the length difference between the trait generic length and the impl generic length.  I would have expected the difference between the trait method generic length and the impl method generic length, since the opaques are based on them in `generics_of`.  We should perhaps explain that this is ok since we check the number of generics before reaching this point.

r=me with that comment